[feature/share-sheet] Import files from share sheet / Create Folder in Directorypicker

hosy opened the pull request:
## Description
User can import files using the iOS share sheet. all file types a…


## Related Issue
#76 
#443 

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

hosy commented:
Would be great if you can give me some feedback, which version you like. This screen will be shown, if `Open In` was triggered from an other app to import a file into the ownCloud app and if the user has more than one account configured.
The user will see this view, after the user will kicked into our app.

In my opinion the CardView is more custom and gives more informations but the AlertView is more clean and more a view to decide something. What do you think?

| CardView                                                                                                                                                         | AlertView                                                                                                                                                        |
|------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ![Simulator Screen Shot - iPhone X - 2019-07-09 at 22 35 40](https://user-images.githubusercontent.com/736109/60922746-b4e54400-a29d-11e9-9a30-643e4ccfc318.png) | ![Simulator Screen Shot - iPhone X - 2019-07-09 at 22 58 24](https://user-images.githubusercontent.com/736109/60922760-c595ba00-a29d-11e9-8e03-877cdb32a1e0.png) |

michaelstingl commented:
Alert view works for me. I‘d tend to the simple solution, then iterate based on user feedback.

mneuwert commented:
i would vote for the card view but if alert is simpler then we could use it and then go for a new iOS13 style context menu as soon as we start working on iOS13 related enhancements

hosy commented:
@michaelstingl @mneuwert both solutions are implemented and can be tested via bitrise build

hosy commented:
Changed the CardView and added a description row and also changed the title:
![Simulator Screen Shot - iPhone X - 2019-07-11 at 22 03 15](https://user-images.githubusercontent.com/736109/61081670-11746a80-a428-11e9-9358-7b51878eb160.png)

felix-schwarz commented on ownCloud/Import/ImportFilesController.swift:
```diff
@@ -0,0 +1,154 @@
+	func importItemWithDirectoryPicker(with url : URL, into bookmark: OCBookmark) {
+		OCCoreManager.shared.requestCore(for: bookmark, setup: { (_, _) in
+		}, completionHandler: { (core, error) in
+			if error == nil {
```
Rather than using `core!` a few lines below, please use `if let core = core, error == nil` here for added safety.

felix-schwarz commented on ownCloud/Import/ImportFilesController.swift:
```diff
@@ -0,0 +1,154 @@
+					let pickerNavigationController = ThemeNavigationController(rootViewController: directoryPickerViewController)
+
+					if let wd = UIApplication.shared.delegate?.window {
+						let vc = wd!.rootViewController
```
Please use `window` instead of `wd` and `viewController` instead of `vc` to match the code style of the rest of the code base.

felix-schwarz commented on ownCloud/Import/ImportFilesController.swift:
```diff
@@ -0,0 +1,154 @@
+					if let wd = UIApplication.shared.delegate?.window {
+						let vc = wd!.rootViewController
+						if let navCon = vc as? UINavigationController, let topVC = navCon.visibleViewController {
+							topVC.present(pickerNavigationController, animated: true)
```
Please use `navigationController` instead of `navCon` and `topViewController` instead of `topVC` to match the code style of the rest of the code base.

felix-schwarz commented on ownCloud/Import/ImportFilesController.swift:
```diff
@@ -0,0 +1,154 @@
+			let moreViewController = self.cardViewController(for: url)
+			if let wd = UIApplication.shared.delegate?.window {
+				let vc = wd!.rootViewController
+				if let navCon = vc as? UINavigationController, let topVC = navCon.visibleViewController {
```
Please use `window` instead of `wd`, `viewController` instead of `vc`, `navigationController` instead of `navCon` and `topViewController` instead of `topVC` to match the code style of the rest of the code base.

felix-schwarz commented on ownCloud/Import/ImportFilesController.swift:
```diff
@@ -0,0 +1,154 @@
+					let directoryPickerViewController = ClientDirectoryPickerViewController(core: core!, path: "/", selectButtonTitle: "Save here".localized, avoidConflictsWith: [], choiceHandler: { [weak self] (selectedDirectory) in
+
+						if let targetDirectory = selectedDirectory {
+							self?.importFile(url: url, to: targetDirectory, core: core)
```
`importFile` is never called here because there's `[weak self]` for this block a few lines above and the `ImportFilesController` is released before this block is run (testing with a single account setup, YMMV).

Please make `self` non-weak here.

felix-schwarz commented on ownCloud/Import/ImportFilesController.swift:
```diff
@@ -0,0 +1,154 @@
+	}
+
+	func importItemWithDirectoryPicker(with url : URL, into bookmark: OCBookmark) {
+		OCCoreManager.shared.requestCore(for: bookmark, setup: { (_, _) in
```
The `requestCore` for the `OCCore` needs to be balanced with a `returnCore` in another place.

The best place for this would be in the `placeholderCompletionHandler` passed to `core.importFileNamed(…)` - with a short delay of 1 or 2 seconds to give `OCCore` a chance to schedule the background `NSURLSessionTask` to upload the file.

Putting it in the `completionHandler` passed to `core.importFileNamed(…)`, on the other hand, would guarantee that the core stays open until the upload has finished. However, if the user doesn't have connectivity, that could keep the core open indefinitely.

hosy replied:
Please can you share a code example?

felix-schwarz replied:
Sure!

```swift
		if let progress = core?.importFileNamed(name,
							at: targetDirectory,
							from: url,
							isSecurityScoped: false,
							options: [OCCoreOption.importByCopying : true,
								  OCCoreOption.automaticConflictResolutionNameStyle : OCCoreDuplicateNameStyle.bracketed.rawValue],
							placeholderCompletionHandler: { (error, item) in
								if error != nil {
									Log.debug("Error uploading \(Log.mask(name)) to \(Log.mask(targetDirectory.path)), error: \(error?.localizedDescription ?? "" )")
								}

								// Return OCCore
								OCCoreManager.shared.returnCore(for: bookmark, completionHandler: nil)
							},
							resultHandler: { (error, _ core, _ item, _) in
…
```

felix-schwarz commented on ownCloud/Import/ImportFilesController.swift:
```diff
@@ -0,0 +1,154 @@
+
+														let fileManager = FileManager.default
+														do {
+															try fileManager.removeItem(at: url)
```
If `true` was passed for [`UIApplicationOpenURLOptionsOpenInPlaceKey`](https://developer.apple.com/documentation/uikit/uiapplicationopenurloptionsopeninplacekey?language=objc), this would delete the original file. Please make sure to honor the value passed for [`UIApplicationOpenURLOptionsOpenInPlaceKey`](https://developer.apple.com/documentation/uikit/uiapplicationopenurloptionsopeninplacekey?language=objc) before deleting the file (possibly by using the approach I outlined in the review comment in `AppDelegate.swift`).

felix-schwarz commented on ownCloud/Resources/en.lproj/Localizable.strings:
```diff
@@ -406,3 +406,7 @@
+
+/* Import File */
+"Save File" = "Save File";
+"Select an account where to import the file and choose a destination directory." = "Select an account where to import the file and choose a destination directory.";
```
Just an alternative suggestion: `Choose an account and directory to import the file into.`

hosy replied:
Used one of your description

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/ios-app?pullRequest=445) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/ios-app?pullRequest=445) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: hosy<br/>:x: felix-schwarz<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/ios-app?pullRequest=445) it.</sub>

jesmrec commented:
### (1)

I am able to share content from the Safari browser, for example a PDF file. But, in Photos app, the ownCloud app does not appear to share with. do i miss something? 

Safari:
![Screenshot 2019-07-24 at 16 35 14](https://user-images.githubusercontent.com/14894746/61802603-0d394b80-ae31-11e9-9f7b-dfe47a10737e.png)

Photos app:

![Jul-24-2019 16-37-07](https://user-images.githubusercontent.com/14894746/61802762-538eaa80-ae31-11e9-99b2-cfe9f15f8105.gif)

of course, both screenshots were taken with the same device with the ownCloud app installed.

jesmrec commented:
only as a tip, out of scope of this issue: If a user has two accounts in the same server (f.ex., an user account and an admin account) with the same full name, accounts can not be distinguished:

![Screenshot 2019-07-24 at 16 56 59](https://user-images.githubusercontent.com/14894746/61804355-14158d80-ae34-11e9-91ec-d1ceb9dc13ab.png)

this is out of scope because the same issue happens in the account list. i will open a separate issue about that.

jesmrec commented:
### (2)

1. Select several files to share from an external app
2. Select an account to upload to

Current: only one file is uploaded
Expected: all files are uploaded

![Jul-24-2019 17-28-39](https://user-images.githubusercontent.com/14894746/61806794-9d2ec380-ae38-11e9-8459-8ac039c837ae.gif)

michaelstingl commented:
> ### (1)
> I am able to share content from the Safari browser, for example a PDF file, Pages App, Files App... but, in Photos app, the ownCloud app does not appear to share with. do i miss something?

Yeah, I think iOS behaves different in some apps. We'd need to add a FAQ section to the docs and open a new issue for "Share Sheet v2"

mneuwert commented:
### Re (2)
This is a limitation of opening URLs in the app, the `AppDelegate.application(app, url, options)` method which is called in this case will get a single file URL even if multiple are selected. I guess we'll need to implement full blown sharing extension to get rid of this annoyance.

mneuwert commented:
### Re (1)
I tried to extend `LSItemContentTypes` key in Info.plist with more file types (public.image etc) but it had no effect on Photos app. Probably it is Apple's intention to not allow opening image url directly in another app.

jesmrec commented:
(1) :then, this is a [WONT FIX]

(2):

> I guess we'll need to implement full blown sharing extension to get rid of this annoyance.

inside the current PR? or separately. In case this will be addressed in another issue, we should take in account the limitation (CC @michaelstingl)

michaelstingl commented:
> inside the current PR? or separately. In case this will be addressed in another issue, we should take in account the limitation (CC @michaelstingl)

No, new PR. I'll give the Share Sheet v1 a try. Probably it's better than no share sheet. Then work on Share Sheet v2 …

michaelstingl commented:
While we still use the limited v1-Share-Sheet, we could help users to discover the "Save to Files" entry in the Share Sheet…

jesmrec commented:
### (3)

The icon to share with contains a different name in different apps, f. ex:

- ownCloud: "Copy in ownCloud"
- iCloud: "Open to ownCloud"

i guess it is not relevant, but it took my attention.

jesmrec commented:
### (4)

Having problems to share form Files App. In terms of UI, after sharing from Files App, the following log:

```
2019-08-01 08:16:41.012000+0200 ownCloud[284.006252] [ERRO] | [CORE, CommandLocalImport] Local creation for item <OCItem: 0x10a24f2a0, type: 0, name: Term Paper.pages, path: /Term Paper.pages, size: 0 bytes, MIME-Type: (null), Last modified: (null), Last used: (null) fileID: _placeholder_8E3A6DB1-21F9-41F0-9C19-D7C908C114E4, eTag: _placeholder_, parentID: 00002598ocfizvgliqbl, localID: 911A74292B8D412CB2D9B2BA1ADAB554, parentLocalID: D972724BF5454EAF9969FFE7A37CE767> from file:///private/var/mobile/Library/Mobile%20Documents/com~apple~Pages/Documents/Term%20Paper.pages to file:///private/var/mobile/Containers/Shared/AppGroup/B2AD5B71-5A1F-4A41-9B2F-2CDB4579DBEA/File%20Provider%20Storage/886C0A1F-576F-48E6-925C-E0CDAE8379ED/911A74292B8D412CB2D9B2BA1ADAB554/Term%20Paper.pages failed in move phase with error: Error Domain=NSCocoaErrorDomain Code=257 "The file “Term Paper.pages” couldn’t be opened because you don’t have permission to view it." UserInfo={NSFilePath=/private/var/mobile/Library/Mobile Documents/com~apple~Pages/Documents/Term Paper.pages, NSUnderlyingError=0x10a227ae0 {Error Domain=NSPOSIXErrorDomain Code=1 "Operation not permitted"}} [OCCore+CommandLocalImport.m:139|FULL]
2019-08-01 08:16:41.021000+0200 ownCloud[284.006252] [dbug] | [APP] Error uploading Term Paper.pages to /, error: The file “Term Paper.pages” couldn’t be opened because you don’t have permission to view it. [ImportFilesController.swift:149|FULL]
2019-08-01 08:16:41.027000+0200 ownCloud[284.006252] [dbug] | [APP] Error uploading Term Paper.pages to /, error: The file “Term Paper.pages” couldn’t be opened because you don’t have permission to view it. [ImportFilesController.swift:159|FULL]
2019-08-01 08:16:41.028000+0200 ownCloud[284.006252] [dbug] | [APP] Error setting up upload of Term Paper.pages to / [ImportFilesController.swift:167|FULL]
```

felix-schwarz commented:
@jesmrec #471 fixes finding (4), along with two other issues (described in the PR).

settermjd commented:
I can't document this functionality, as it is not working in version 1.1 of the app, which I have for testing.

michaelstingl commented:
> I can't document this functionality, as it is not working in version 1.1 of the app, which I have for testing.

Please open a new issue and describe the problem.
